homeassistant/components/mqtt/__init__.py

```diff
@@ -465,18 +466,24 @@ def __init__(self, hass, broker, port, client_id, keepalive, username,
             if tls_insecure is not None:
                 self._mqttc.tls_insecure_set(tls_insecure)
 
-        self._mqttc.on_subscribe = self._mqtt_on_subscribe
-        self._mqttc.on_unsubscribe = self._mqtt_on_unsubscribe
-        self._mqttc.on_connect = self._mqtt_on_connect
-        self._mqttc.on_disconnect = self._mqtt_on_disconnect
-        self._mqttc.on_message = self._mqtt_on_message
+        self._mqttc.on_subscribe = self._safe_cb(self._mqtt_on_subscribe)
+        self._mqttc.on_unsubscribe = self._safe_cb(self._mqtt_on_unsubscribe)
+        self._mqttc.on_connect = self._safe_cb(self._mqtt_on_connect)
+        self._mqttc.on_disconnect = self._safe_cb(self._mqtt_on_disconnect)
+        self._mqttc.on_message = self._safe_cb(self._mqtt_on_message)
 
         if will_message:
             self._mqttc.will_set(will_message.get(ATTR_TOPIC),
                                  will_message.get(ATTR_PAYLOAD),
                                  will_message.get(ATTR_QOS),
                                  will_message.get(ATTR_RETAIN))
 
+    def _safe_cb(self, callback_func):
```

Comment:
I don't think that is need. paho run inside own thread and we are safe to call that function.

Reply:
I think this is one of the original problems, exactly because paho is running in a [different thread](https://github.com/eclipse/paho.mqtt.python/blob/e9914a759f9f5b8081d59fd65edfd18d229a399e/src/paho/mqtt/client.py#L1513) from hass the main loop.

Without adding the callbacks to the event loop, there's the possibility of a call to the `on_subscribe` callback to happen while the `async_subscribe` call is still being executed. This is problematic because the callback might finish executing even before the`progress` map is updated with the generated `mid` and the `topic` name here:

https://github.com/home-assistant/home-assistant/blob/c191c13f3acb52126c77872f17e77d4107817cf0/homeassistant/components/mqtt/__init__.py#L536
Causing the `subscribed_topics` map never to be updated here:
https://github.com/home-assistant/home-assistant/blob/c191c13f3acb52126c77872f17e77d4107817cf0/homeassistant/components/mqtt/__init__.py#L582-L586

The approach I took by delaying callback calls and adding then to event loop is somewhat similar to what https://github.com/mossblaser/aiomqtt lib does. aiomqtt is an asyncio safe wrapper to paho.mqtt, I even considered replacing paho.mqtt entirely with aiomqtt but considered it a change too big to be worth it.

Reply:
`_mqtt_*` need to run in a diferent thread like the main loop. And that is with your variant and with the variant they we have current. So it's not needed. We use the dispatcher/add_job function to change beteen loop/thread

Reply:
Pascal has a good point. Let's remove the usage of `_save_cb` for now. If the problem persists we can explore adding them back.

Reply:
I think @cezarsa's point is that all MQTT callbacks (`_mqtt_*`) and all methods like `async_subscribe` should run on the _same_ thread as otherwise you need synchronization around shared data structures to avoid race conditions.

Using `add_job` from within the internal MQTT hread will queue the callback so it will be executed in the hass asyncio event loop and since `async_publish` is also executed within the same event loop there is no possibility for synchronization issues.

Reply:
I actually added some logging to this fix, and it seems that the actual race condition is caused by the fact that the _mqtt_on_connect callback is for some reason called twice in parallel by different threads.

```
2017-11-06 17:38:13 ERROR (Thread-7) [homeassistant.components.mqtt] safe_cb: _mqtt_on_connect
2017-11-06 17:38:14 ERROR (Thread-2) [homeassistant.components.mqtt] safe_cb: _mqtt_on_subscribe
2017-11-06 17:38:14 ERROR (Thread-6) [homeassistant.components.mqtt] safe_cb: _mqtt_on_message
2017-11-06 17:38:38 ERROR (Thread-8) [homeassistant.components.mqtt] safe_cb: _mqtt_on_message
...
<disconnected network cable>
...
2017-11-06 17:40:14 ERROR (Thread-10) [homeassistant.components.mqtt] safe_cb: _mqtt_on_disconnect
2017-11-06 17:40:14 ERROR (Thread-3) [homeassistant.components.mqtt] safe_cb: _mqtt_on_disconnect

2017-11-06 17:40:14 WARNING (Thread-4) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 1 s
2017-11-06 17:40:14 WARNING (Thread-9) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 1 s
2017-11-06 17:40:15 WARNING (Thread-4) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 2 s
2017-11-06 17:40:15 WARNING (Thread-9) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 2 s
2017-11-06 17:40:17 WARNING (Thread-4) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 4 s
2017-11-06 17:40:17 WARNING (Thread-9) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 4 s
2017-11-06 17:40:21 WARNING (Thread-9) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 8 s
2017-11-06 17:40:21 WARNING (Thread-4) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 8 s
2017-11-06 17:40:29 WARNING (Thread-4) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 16 s
2017-11-06 17:40:29 WARNING (Thread-9) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 16 s
2017-11-06 17:40:45 WARNING (Thread-4) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 32 s
2017-11-06 17:40:45 WARNING (Thread-9) [homeassistant.components.mqtt] Disconnected from MQTT (1). Trying to reconnect in 32 s
2017-11-06 17:41:17 ERROR (Thread-3) [homeassistant.components.mqtt] safe_cb: _mqtt_on_connect
2017-11-06 17:41:17 ERROR (Thread-6) [homeassistant.components.mqtt] safe_cb: _mqtt_on_subscribe
2017-11-06 17:41:17 ERROR (Thread-5) [homeassistant.components.mqtt] safe_cb: _mqtt_on_message
2017-11-06 17:41:17 ERROR (Thread-3) [homeassistant.components.mqtt] safe_cb: _mqtt_on_connect
2017-11-06 17:41:17 ERROR (Thread-12) [homeassistant.components.mqtt] safe_cb: _mqtt_on_subscribe
2017-11-06 17:41:17 ERROR (Thread-5) [homeassistant.components.mqtt] safe_cb: _mqtt_on_message
2017-11-06 17:41:18 ERROR (Thread-4) [homeassistant.components.mqtt] safe_cb: _mqtt_on_message
```

```diff
diff --git a/homeassistant/components/mqtt/__init__.py b/homeassistant/components/m
index 1da431d..c6fe441 100644
--- a/homeassistant/components/mqtt/__init__.py
+++ b/homeassistant/components/mqtt/__init__.py
@@ -478,11 +478,18 @@ class MQTT(object):
                                  will_message.get(ATTR_QOS),
                                  will_message.get(ATTR_RETAIN))
 
+
     def _safe_cb(self, callback_func):
         # paho.mqtt callbacks run in another thread, here we add the callback
         # task in the event loop thread avoiding possible synchronization
         # problems.
-        return lambda *args: self.hass.add_job(callback_func, *args)
+
+        def logging_callback(*args):
+            _LOGGER.error("safe_cb: %s" % callback_func.__name__)
+            self.hass.add_job(callback_func, *args)
+
+        #return lambda *args: self.hass.add_job(callback_func, *args)
+        return lambda *args: self.hass.add_job(logging_callback, *args)
 
     @asyncio.coroutine
     def async_publish(self, topic, payload, qos, retain):

```

So even though this PR fixes the actual race and it IMO makes sense to keep the safe_cb protection, it's still a bit broken.

Reply:
> I actually added some logging to this fix, and it seems that the actual race condition is caused by the fact that the _mqtt_on_connect callback is for some reason called twice in parallel by different threads.

Hm, I checked again and it seems that the callback is being using `run_in_executor` which dispatches to the executor thread pool. I think this is wrong, and these callbacks should be changed to be coroutines so they will be instead dispatched inside the event loop directly.

Reply:
Actually the dual call to the disconnect callback above seems to be bug in paho:
https://github.com/eclipse/paho.mqtt.python/issues/246

Reply:
That may be, but as long as the above modifications are made, they will never be called in parallel with each other or in parallel with other coroutines running in the event loop thread.

Reply:
If the issue is in paho, can we remove the add callback feature? 

Because with the CB they will run in the worker thread in parallel. They will not be run on the event loop.

Reply:
With the current implementation, all CB run in same thread (that from paho loop). With this implementation it run inside threadpool pararell and that is what the initial post not will do and now do with that PR.

If we migrate the callback to async to run inside loop, we have also no control over the correct running sequent while that is handle by asyncio. With current solution we have the best sync.

Reply:
> If the issue is in paho, can we remove the add callback feature?

If this is the only issue and there is actually no race condition, then this sounds ok. But if @cezarsa is correct and there is actually a race condition, then something must be done about it, either by introducing locks or ensuring that specific code does not run in parallel (e.g. by scheduling it in the asyncio event loop).

> Because with the CB they will run in the worker thread in parallel. They will not be run on the event loop.

This is why I mentioned that it should use the callback decorator so it will be scheduled in the event loop.

> If we migrate the callback to async to run inside loop, we have also no control over the correct running sequent while that is handle by asyncio. With current solution we have the best sync.

If you schedule the callback in the asyncio event loop, you have the serialization guarantee (e.g., if you don't yield, the callback cannot be interrupted so there are no race conditions). The main issue is that the code should not run in parallel as it could corrupt state.

> We are faster for process mqtt if you do it without handle all inside loop. So we are handling only relevent mqtt stuff inside our event loop and not the hole traffic.

You don't need to forward everything. The problem is only with code, which manipulates shared data structures. That code should be either protected by locks or it should be ensured that the code does not execute in parallel (as mentioned above).

Reply:
@cezarsa it seems that it is difficult to get an agreement on if the _safe_cb() function is the right thing to do or not, but the rest of the code will be generally beneficial. Would you please mind to split this PR into two parts so while we will keep us busy discussing the race condition, the other part could be accepted right away ? The MQTT issues above are quite painful for me so I'd like to see at least partial relief sooner rather than later.
Or if you are busy I'll be more than happy to do the split on your behalf (once I get your blessing).

Thanks a lot.

Reply:
Sorry for not joining the discussion sooner, I've been quite busy the last few weeks and didn't check this. So, I've just update the code removing the calls to _safe_cb() for now so this can be merged.

I still think some kind of synchronization is needed to ensure correctness though, but I'm not able to prove it right now. Initially I intended on writing a test case simulating the behavior of paho-mqtt when calling callbacks and stressing the subscribe workflow to check if this would actually lead to inconsistencies, however I haven't had time to do that. I was able to force inconsistencies (qos never added to subscribed_topics even though the topic was subscribed) adding a small `time.sleep()` right after the `_mqttc.subscribe` yield call but this isn't really critical anymore as wanted_callback is the one used in re-subscriptions.

If I find some time in the future I'll try coming back to this with a test based approach.